Events and logs api (#2676)
open-telemetry/opentelemetry-specification: specification/logs/api.md

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [LoggerProvider](#loggerprovider)
  * [LoggerProvider operations](#loggerprovider-operations)
    + [Get a Logger](#get-a-logger)
- [Logger](#logger)
  * [Logger operations](#logger-operations)

<!-- tocstop -->

The Events and Logs API consist of these main classes:

* LoggerProvider is the entry point of the API. It provides access to Loggers.
* Logger is the class responsible for creating events and logs using Log records.

LoggerProvider/Logger are analogous to TracerProvider/Tracer.

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B --> C(Event)
    B --> D(Log)
```

## LoggerProvider

`Logger`s can be accessed with a `LoggerProvider`.

In implementations of the API, the LoggerProvider is expected to be the stateful
object that holds any configuration.

Normally, the LoggerProvider is expected to be accessed from a central place.
Thus, the API SHOULD provide a way to set/register and access a global default
LoggerProvider.

Notwithstanding any global LoggerProvider, some applications may want to or have
to use multiple LoggerProvider instances, e.g. to have different configuration

> (like LogRecordProcessors) for each (and consequently for the Loggers obtained

Reviewer: Let's say "LogProcessors" since that's what they're called in the [SDK document](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/logging-library-sdk.md#logprocessor). If we want to rename to "LogRecordProcessor", we can do so in a followup PR and update the reference here as well.

Author: Changed to LogProcessors.

work with an outdated configuration or by ensuring that new configuration
applies also to previously returned Loggers.

Note: This could, for example, be implemented by storing any mutable
configuration in the LoggerProvider and having Logger implementation objects
have a reference to the LoggerProvider from which they were obtained.
If configuration must be stored per-Logger (such as disabling a certain Logger),
the Logger could, for example, do a look-up with its name+version+schema_url+event_domain+include_trace_context+attributes
in a map in the LoggerProvider, or the LoggerProvider could maintain a registry
of all returned Loggers and actively update their configuration if it changes.

The effect of associating a Schema URL with a Logger MUST be that the telemetry
emitted using the Logger will be associated with the Schema URL, provided that
the emitted data format is capable of representing such association.

## Logger

The Logger is responsible for creating Events and Logs.

Note that Loggers should not be responsible for configuration. This should be
the responsibility of the LoggerProvider instead.

### Logger operations

The Logger MUST provide functions to:

> - Create a `LogRecord`, representing an `Event` and emit it to the processing

Reviewer 2: Should we define operation names for these operations? It would be beneficial to have such a name definition in the spec to give language implementation a reference how to name their methods.

Reviewer 3: I believe it was intentionally left unspecified, so that different SDKs implement what makes sense based on patterns in their language. Java, for example, would use the builder pattern. In [this JS draft implementation](https://github.com/open-telemetry/opentelemetry-js/pull/3117), we settled on `emitLogRecord(LogRecord)` and `emitEvent(Event)` (see [this conversation](https://github.com/open-telemetry/opentelemetry-js/pull/3117#discussion_r938046782)).

Author: @legendecas the last bullet here does suggest the operation name to use.

pipeline.
  - The API MUST accept an event name as a parameter. The event name provided
  should be recorded as an attribute with key `event.name`. Care MUST be taken
  by the implementation to not override or delete this attribute while the
  `Event` is created to preserve its identity.
  - Events require the `event.domain` attribute. The API MUST not allow creating
  an event if the Logger instance doesn't have `event.domain` scope attribute.
  - This function MAY be named `logEvent`.
- Create a `LogRecord` and emit it to the processing pipeline.
  - This function MAY be named `logRecord`.

>   - The intended users of this API is Log Appenders.

Reviewer: #nit:
```suggestion
  - This API is intended for use by Log Appenders, and SHOULD not be used by end users or other instrumentation. 
```

Author: Applied the suggested change.

work with an outdated configuration or by ensuring that new configuration
applies also to previously returned Loggers.

Note: This could, for example, be implemented by storing any mutable
configuration in the LoggerProvider and having Logger implementation objects
have a reference to the LoggerProvider from which they were obtained.
If configuration must be stored per-Logger (such as disabling a certain Logger),
the Logger could, for example, do a look-up with its name+version+schema_url+event_domain+include_trace_context+attributes
in a map in the LoggerProvider, or the LoggerProvider could maintain a registry
of all returned Loggers and actively update their configuration if it changes.

The effect of associating a Schema URL with a Logger MUST be that the telemetry
emitted using the Logger will be associated with the Schema URL, provided that
the emitted data format is capable of representing such association.

## Logger

The Logger is responsible for creating Events and Logs.

Note that Loggers should not be responsible for configuration. This should be
the responsibility of the LoggerProvider instead.

### Logger operations

The Logger MUST provide functions to:

> - Create a `LogRecord`, representing an `Event` and emit it to the processing

Reviewer 4: WDYT about introducing a separate `Event` interface? It might be implemented as a `LogRecord` in the SDK, but the API would be different (as events are conceptually somewhat different from logs).

Note that a separate API could help with the first req:
> Care MUST be taken by the implementation to not override or delete this attribute while the `Event` is created to preserve its identity.

Author: We discussed at length about having separate interfaces for Logs and Events during [this](https://github.com/open-telemetry/oteps/pull/202#issuecomment-1143986000) OTEP.  The differences between the two approaches are captured in [this](https://docs.google.com/document/d/1Uy5L3UThFyc4PelgH51G3OCLFbvzq2xWiNDy2G8iXxU/edit#) document, and a decision was taken to use one common API interface for Logs and Events. The rationale for the decision is in [this](https://cloud-native.slack.com/archives/C01N5UCHTEH/p1654711359579599) slack thread. 

Further, I would like to understand how a separate API could help with the first req. Since both the event's `name` and event's `attributes` end up in the Log data-model's `attributes` field, this requirement will remain even with a separate API interface for Events.

pipeline.
  - The API MUST accept an event name as a parameter. The event name provided
  should be recorded as an attribute with key `event.name`. Care MUST be taken
  by the implementation to not override or delete this attribute while the
  `Event` is created to preserve its identity.
  - Events require the `event.domain` attribute. The API MUST not allow creating

>   an event if the Logger instance doesn't have `event.domain` scope attribute.

Reviewer 4: Instead of having a runtime check for this, there could be different APIs for creating logs and events: e.g. `LoggerProvider` could allow to create either a `Logger`, or an `EventEmitter` (`EventBuilder`? `EventPublisher`?). WDYT?

Author: Please see the [previous](https://github.com/open-telemetry/opentelemetry-specification/pull/2676#discussion_r951238626) comment.

  - This function MAY be named `logEvent`.

> - Create a `LogRecord` and emit it to the processing pipeline.

Reviewer 4: Isn't the `LogRecord` currently an SDK concept? If we want to use it here, I believe that the API spec should define all the methods/fields of the log object.

Author: I added some text about `LogRecord` data model at the top now.

should be added as `event.domain` attribute of the instrumentation scope.
- `include_trace_context` (optional): Specifies whether the Trace Context should
automatically be passed on to the events and logs created by the Logger. This
SHOULD be true by default.
- `attributes` (optional): Specifies the instrumentation scope attributes to
associate with emitted telemetry.

Implementations MUST return different `Logger` instances when called repeatedly
with different values of parameters. Note that always returning a new `Logger`
instance is a valid implementation. The only exception to this rule is the no-op
`Logger`: implementations MAY return the same instance regardless of parameter
values.

Implementations MUST NOT require users to repeatedly obtain an Logger again with
the same name+version+schema_url+event_domain+include_trace_context+attributes
to pick up configuration changes. This can be achieved either by allowing to
work with an outdated configuration or by ensuring that new configuration
applies also to previously returned Loggers.

Note: This could, for example, be implemented by storing any mutable

> configuration in the LoggerProvider and having Logger implementation objects

Reviewer 5: ```suggestion
configuration in the `LoggerProvider` and having `Logger` implementation objects
```

> have a reference to the LoggerProvider from which they were obtained.

Reviewer 5: Let's use formatting to what we do in trace API:
```suggestion
have a reference to the `LoggerProvider` from which they were obtained.
```

> If configuration must be stored per-Logger (such as disabling a certain Logger),

Reviewer 5: ```suggestion
If configuration must be stored per-Logger (such as disabling a certain `Logger`),
```

> the Logger could, for example, do a look-up with its name+version+schema_url+event_domain+include_trace_context+attributes

Reviewer 5: ```suggestion
the `Logger` could, for example, do a look-up with its name+version+schema_url+event_domain+include_trace_context+attributes
```

> in a map in the LoggerProvider, or the LoggerProvider could maintain a registry

Reviewer 5: ```suggestion
in a map in the `LoggerProvider`, or the `LoggerProvider` could maintain a registry
```

> of all returned Loggers and actively update their configuration if it changes.

Reviewer 5: ```suggestion
of all returned `Logger`s and actively update their configuration if it changes.
```

> The effect of associating a Schema URL with a Logger MUST be that the telemetry

Reviewer 5: ```suggestion
The effect of associating a Schema URL with a `Logger` MUST be that the telemetry
```

> emitted using the Logger will be associated with the Schema URL, provided that

Reviewer 5: ```suggestion
emitted using the `Logger` will be associated with the Schema URL, provided that
```

the emitted data format is capable of representing such association.

## Logger

> The Logger is responsible for creating Events and Logs.

Reviewer 5: ```suggestion
The `Logger` is responsible for creating Events and Logs.
```

> Note that Loggers should not be responsible for configuration. This should be

Reviewer 5: ```suggestion
Note that `Logger`s should not be responsible for configuration. This should be
```

> the responsibility of the LoggerProvider instead.

Reviewer 5: ```suggestion
the responsibility of the `LoggerProvider` instead.
```

emitted using the Logger will be associated with the Schema URL, provided that
the emitted data format is capable of representing such association.

## Logger

The Logger is responsible for creating Events and Logs.

Note that Loggers should not be responsible for configuration. This should be
the responsibility of the LoggerProvider instead.

### Logger operations

The Logger MUST provide functions to:

- Create a `LogRecord`, representing an `Event` and emit it to the processing
pipeline.
  - The API MUST accept an event name as a parameter. The event name provided
  should be recorded as an attribute with key `event.name`. Care MUST be taken
  by the implementation to not override or delete this attribute while the
  `Event` is created to preserve its identity.
  - Events require the `event.domain` attribute. The API MUST not allow creating
  an event if the Logger instance doesn't have `event.domain` scope attribute.
  - This function MAY be named `logEvent`.
- Create a `LogRecord` and emit it to the processing pipeline.
  - This function MAY be named `logRecord`.
  - This API is intended for use by Log Appenders, and SHOULD not be used by end

>   users or other instrumentation.

Reviewer 6: Isn't this already solved by the `LogEmitter` part of SDK spec? Is there a well-defined use case for why this should be part of API?

Reviewer: The goal is to adjust logging to match the API / SDK layout of the other signals. The [logging-library-sdk](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/logging-library-sdk.md) would transition into the SDK implementation of the API described here. Like the other signals, application owners configure the SDK to decide how logs are processed.

Reviewer 5: Discussed this in the Log SIG call. The current API discourages using logging API as a general-purpose logging API in application code. The API is designed for specific use cases (logging library appender and structured event emitting).
The reason it is now part of the API and not SDK is to align better and be more consistent with how Tracing/Metrics API and SDK are structured.

> ## Events and Logs API Interface

Reviewer 7: ```suggestion
# Events and Logs API Interface
```

**Status**: [Experimental](../document-status.md)

> 

Reviewer 7: Not sure if you are still working on it, but we probably need a TOC here

```suggestion
<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->
<!-- toc -->

<!-- tocstop -->

```

Author: TOC added.

The Events and Logs API consist of these main classes:

* LoggerProvider is the entry point of the API. It provides access to Loggers.
* Logger is the class responsible for creating events and logs using Log records.

LoggerProvider/Logger are analogous to TracerProvider/Tracer.

> ![Events and Logs API classes](img/events-and-logs-api.png)

Reviewer 8: Could this be changed to a Mermaid diagram? e.g. https://github.com/open-telemetry/opentelemetry-demo/blob/main/docs/v1Graph.md

### LoggerProvider

Logger can be accessed with an LoggerProvider.

> In implementations of the API, the LoggerProvider is expected to be the stateful object that holds any configuration. (Note: The SDK implementation of this is what we currently call the [LogEmitterProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/logging-library-sdk.md#logemitterprovider))

Reviewer 7: Doesn't seem it's enforced via markdownlint, but I usually don't see long lines in the markdown files in the spec repo. Most seem to fall under the 80 character length.

Author: Modified the text to limit the lines to 80 characters length.
Remove the 'Note' as well - the SDK spec will be updated to match the API separately.

Normally, the LoggerProvider is expected to be accessed from a central place. Thus, the API SHOULD provide a way to set/register and access a global default LoggerProvider.

Notwithstanding any global LoggerProvider, some applications may want to or have to use multiple LoggerProvider instances, e.g. to have different configuration (like LogRecordProcessors) for each (and consequently for the Loggers obtained from them), or because it's easier with dependency injection frameworks. Thus, implementations of LoggerProvider SHOULD allow creating an arbitrary number of instances.

#### LoggerProvider operations

The LoggerProvider MUST provide the following functions:

> * Get an Logger

Reviewer 9: ```suggestion
* Get a Logger
```

> ##### Get an Logger

Reviewer 9: ```suggestion
##### Get a Logger
```

- `version` (optional): Specifies the version of the instrumentation scope if the scope has a version (e.g. a library version). Example value: 1.0.0.
- `schema_url` (optional): Specifies the Schema URL that should be recorded in the emitted telemetry
- `event_domain` (optional): Specifies the domain for the events created, which should be added in the attribute `event.domain` in the instrumentation scope.

> - `include_trace_context` (optional): Specifies whether the Trace Context should automatically be passed on to the events and logs created by the Logger. This SHOULD be false by default.

Reviewer 10: A few questions here:

- Why is the default false?  OTEL's mission is correlated telemetry, we do see trace context-in-logs being super valuable for users.  
- Should we also have a mechanism that allows BAGGAGE to be included here?
- Isn't attaching trace context an SDK concern, not an API concern?

Generally, I think the interaction of Baggage w/ signals is far to user-intensive right now, this is a great example where it makes a lot of sense to append baggage attributes as a configuration parameter.

However, at a higher level this kind of associativity (e.g. metric exemplars) are controlled at the SDK level.  E.g. we DON'T want users have to change instrumentation to get better correlated telemetry.

Author: There are situations when the event handler callbacks are called on the same thread that is executing a span and the event handler could be unrelated the span in progress. it is therefore the instrumentation developer's concern whether to associate the trace context of the span in progress with any events and logs created from the event handler. The SDK cannot make this decision and hence this is exposed as a flag in the API. I chose the default to be false so that there is no chance of accidental association which could be misleading.

I am not aware of any use-case/requirement to include Baggage in the events/logs.

Author: Ok, I will change the default for `include_trace_context`.

Author: @jsuereth can I resolve this conversation?
